hedera-node/docs/design/services/smart-contract-service/IHFC-associate-dissociate.md

```diff
@@ -0,0 +1,81 @@
+# IHRC Facade Associate and Dissociate Functionality
+
+## Purpose
+
+[HIP-218](https://hips.hedera.com/hip/hip-218) proposed a way to enable treating HTS tokens in an equivalent manner
+to ERC tokens in the EVM world by creating a proxy redirect facade contract.  We need to extend 
+this functionality to include `associate` and `dissociate` functions which are part of functionality available to HTS tokens
+but not ERC tokens.
+
+## Goals
+
+- Extend the proxy redirect token facade contract to enable calling `associate` and `dissociate` functions on tokens.
+
+## Non Goals
+
+- Refactor the current implementation to better support future additions to the token facade contract functionality.
+
+## Architecture
+
+The method in which the redirection to proxy contract is implemented is that during the execution of an EVM request, we examine the contract address to see
+if it is an HTS token address and if so we redirect the request to the proxy contract.  This mechanism will be extended to include
+calls to `associate` and `dissociate` functions in the HTSPrecompileContact class method which handles ABI_ID_REDIRECT_FOR_TOKEN function selector.
+
+The following table describes the function selector for the new `associate` and `dissociate` functions and the associated function signatures.
+
+| Function Selector | Function Signature |
+|-------------------|--------------------|
+| 0x0a754de6        | associate()        |
+| 0x5c9217e0        | dissociate()       |
+
+No arguments are necessary because 
+- The token address was already determined by looking up the contract address to determine if it is an HTS token address.
+- The address to associate/dissociate will be the caller (i.e. msg.sender)
+
+Once the above functionality has been implemented in services, the end user will be able to call the `associate` and `dissociate` functions as follows:
+
+```
+IHRC(tokenAddress).associate()
+IHRC(tokenAddress).dissociate()
+```
+
+The solidity interface for IHRC will be the following
+
+```
+interface IHRC {
+    function associate() external returns (responseCode);
+    function dissociate() external returns (responseCode);
+}
+```
+
+## Implementation
+
+Override the existing `AbstractAssociatePrecompile` and `AbstractDissociatePrecompile` classes to handle this new use case
+of being called via the proxy contract.  The new overwritten classes will differ from the existing classes in that they will
+be passed the token id and the sender address as constructor arguments and construct the associate and dissociate operator 
+classes from these arguments.  It will also need to send a value into the constructor of the parent class to signal that
+signature checking will not be required. 
```

Comment:
This differs somewhat from the precompile execution flow we have been following up until this point where the input bytes received from the EVM are decoded when the `body` method of each precompile subclass is called. 
I think you will need to decode the input bytes in order to find out which token address should we be working on. It is the first (and in this case only) argument for the `associate` and `dissociate` functions, inserted into the argument list by the redirect bytecode contract's fallback function.
A synthetic transaction is also created at this point, which is later used to externalize the precompile call result in the record stream.

Reply:
I believe that the tokenId has already been determined [here](https://github.com/hashgraph/hedera-services/blob/29e49604eff059c6bc0c0a4dd2a738f194b32c04/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/HTSPrecompiledContract.java#L563) and I am merely passing the tokenId and the sender address into the constructor and use these values to create and populate the `associateOp` and `dissociateOp` variables instead of having the decode the input bytes.